src/core/radio/ble_secure.cpp

```diff
@@ -0,0 +1,475 @@
+                    Flush();
+
+                    if (error == kErrorAbort)
+                    {
+                        Disconnect(); // Keep BLE active (for testing)
+                        // Stop();        // Stop BLE interface
+                        ExitNow();
+                    }
+                }
+            }
+            else
+            {
+                mReceivedMessage->SetOffset(offset);
+                mReceiveCallback.InvokeIfSet(&GetInstance(), mReceivedMessage, OT_TCAT_MESSAGE_TYPE_NONE, "");
+            }
+
+            SuccessOrExit(mReceivedMessage->SetLength(0)); // also sets the offset to 0
+            requiredBytes = sizeof(Tlv);
+        }
+    }
+    else
+    {
+        SuccessOrExit(mReceivedMessage->AppendBytes(aBuf, aLength));
+
+        mReceiveCallback.InvokeIfSet(&GetInstance(), mReceivedMessage, OT_TCAT_MESSAGE_TYPE_NONE, "");
+        mReceivedMessage->SetLength(0);
+    }
+
+exit:
+    return;
+}
+
+void BleSecure::HandleTransmit(Tasklet &aTasklet)
+{
+    static_cast<BleSecure *>(static_cast<TaskletContext &>(aTasklet).GetContext())->HandleTransmit();
+}
+
+void BleSecure::HandleTransmit(void)
+{
+    Error        error   = kErrorNone;
+    ot::Message *message = mTransmitQueue.GetHead();
+
+    VerifyOrExit(message != nullptr);
+    mTransmitQueue.Dequeue(*message);
+
+    if (mTransmitQueue.GetHead() != nullptr)
+    {
+        mTransmitTask.Post();
+    }
+
+    SuccessOrExit(error = mTls.Send(*message, message->GetLength()));
+
+exit:
+    if (error != kErrorNone)
+    {
+        LogNote("Transmit: %s", ErrorToString(error));
+        message->Free();
+    }
+    else
+    {
+        LogDebg("Transmit: %s", ErrorToString(error));
+    }
+}
+
+Error BleSecure::HandleTransport(void *aContext, ot::Message &aMessage, const Ip6::MessageInfo &aMessageInfo)
+{
+    OT_UNUSED_VARIABLE(aMessageInfo);
+    return static_cast<BleSecure *>(aContext)->HandleTransport(aMessage);
+}
+
+Error BleSecure::HandleTransport(ot::Message &aMessage)
+{
+    otBleRadioPacket packet;
+    uint16_t         len    = aMessage.GetLength();
+    uint16_t         offset = 0;
+    otError          error  = OT_ERROR_NONE;
+
+    while (len > 0)
+    {
+        if (len <= mMtuSize - kGattOverhead)
+            packet.mLength = len;
+        else
+            packet.mLength = mMtuSize - kGattOverhead;
+
+        if (packet.mLength > kPacketBufferSize)
+            packet.mLength = kPacketBufferSize;
+
+        aMessage.Read(offset, mPacketBuffer, packet.mLength);
+        packet.mValue = mPacketBuffer;
+        packet.mPower = 0;
+
+        error = otPlatBleGattServerIndicate(&GetInstance(), kTxBleHandle, &packet);
+        if (error != OT_ERROR_NONE)
+            break;
+
+        len -= packet.mLength;
+        offset += packet.mLength;
+    }
+
+    // if(error != OT_ERROR_NONE)  --> close connection?  Not free message?
+
+    aMessage.Free();
```

Comment:
Do we want to free the message always, or only when Indication was successful?

Reply:
After long discussion in the ThreadGroup workgroup we concluded there is no cure if sending GATT characteristics updates with indicate or notify finally fails after all link layer retries (handled inside the BLE stack). The BLE Core specification considers the connection broken in this case and the BLE stack should close the connection. Therefore it is safe to delete the Openhread message here. Data will remain in the buffer of the BLE stack for possible link layer retries.

Reply:
IHMO this function needs to be refactored.

Reply:
fixed.

Reply:
Because not free'ing the message in the error case is non-obvious, we should perhaps add a comment here why it's not freed in the error case. I think I understand Arnulf's comment but it's not straightforward.  E.g. will the message buffer be automatically freed when the TLS connection is aborted in this case?